src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -81,9 +83,74 @@ var {
   resetContext,
 } = require('ReactFiberContext');
 
+var onCommitRootInDev = null;
+var callCommitListenerInDev = null;
+
 if (__DEV__) {
   var ReactFiberInstrumentation = require('ReactFiberInstrumentation');
   var ReactDebugCurrentFiber = require('ReactDebugCurrentFiber');
+  var warning = require('warning');
+
+  var devCommitListeners = [];
+  var devMountedRoots = new Set();
+
+  callCommitListenerInDev = function(root, listener) {
+    if (__DEV__) {
+      try {
+        listener(root);
+      } catch (err) {
+        warning(
+          false,
+          'React DevTools error: %s.',
+          err,
+        );
+      }
+    }
+  };
+
+  onCommitRootInDev = function(root) {
+    const current = root.current;
+    const isKnownRoot = devMountedRoots.has(root);
+    const isUnmounting = current.memoizedState == null || current.memoizedState.element == null;
+    // Keep track of mounted roots so we can hydrate when DevTools connect.
+    if (!isKnownRoot && !isUnmounting) {
+      devMountedRoots.add(root);
+    } else if (isKnownRoot && isUnmounting) {
+      devMountedRoots.delete(root);
+    }
+    // Notify the listeners.
+    for (let i = 0; i < devCommitListeners.length; i++) {
+      if (callCommitListenerInDev) {
+        callCommitListenerInDev(root, devCommitListeners[i]);
+      }
+    }
+  };
+
+  if (
+    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ !== 'undefined' &&
+    __REACT_DEVTOOLS_GLOBAL_HOOK__.supportsFiber
+  ) {
+    __REACT_DEVTOOLS_GLOBAL_HOOK__.inject({
```

Comment:
Looks like I'll have to move this back to the renderer.
It will need access to the component tree for "jump to component based on DOM node" feature.